format/format.go

```diff
@@ -144,40 +147,46 @@ func AstWithOpts(x interface{}, opts Opts) ([]byte, error) {
 		indent: "\t",
 	}
 
+	var errs ast.Errors = make([]*ast.Error, 0)
+
 	switch x := x.(type) {
 	case *ast.Module:
 		for kw := range extraFutureKeywordImports {
 			x.Imports = ensureFutureKeywordImport(x.Imports, kw)
 		}
-		w.writeModule(x, o)
+		w.writeModule(x, o, &errs)
 	case *ast.Package:
-		w.writePackage(x, nil)
+		w.writePackage(x, nil, &errs)
 	case *ast.Import:
-		w.writeImports([]*ast.Import{x}, nil)
+		w.writeImports([]*ast.Import{x}, nil, &errs)
 	case *ast.Rule:
-		w.writeRule(x, false /* isElse */, o, nil)
+		w.writeRule(x, false /* isElse */, o, nil, &errs)
 	case *ast.Head:
 		w.writeHead(x,
 			false, // isDefault
 			false, // isExpandedConst
 			o,
-			nil)
+			nil,
+			&errs)
 	case ast.Body:
-		w.writeBody(x, nil)
+		w.writeBody(x, nil, &errs)
 	case *ast.Expr:
-		w.writeExpr(x, nil)
+		w.writeExpr(x, nil, &errs)
 	case *ast.With:
-		w.writeWith(x, nil, false)
+		w.writeWith(x, nil, false, &errs)
 	case *ast.Term:
-		w.writeTerm(x, nil)
+		w.writeTerm(x, nil, &errs)
 	case ast.Value:
-		w.writeTerm(&ast.Term{Value: x, Location: &ast.Location{}}, nil)
+		w.writeTerm(&ast.Term{Value: x, Location: &ast.Location{}}, nil, &errs)
 	case *ast.Comment:
-		w.writeComments([]*ast.Comment{x})
+		w.writeComments([]*ast.Comment{x}, &errs)
 	default:
 		return nil, fmt.Errorf("not an ast element: %v", x)
 	}
 
+	if len(errs) > 0 {
```

Comment:
What do you think about adding an `errs` field to `w`? It should be available to all the calls, and we wouldn't need to thread it in. Here, we'd do
```go
if len(w.errs) > 0 {
```